### PR TITLE
clickhouse-benchmark enhancements

### PR DESCRIPTION
**Pull request by ludv1x**
Three new options:
- launches limit
- time limit
- final results in json file

### CODE REVIEW

File: dbms/src/Client/Benchmark.cpp [hunk 1]
```diff
@@ -320,17 +340,60 @@ class Benchmark
...
+		print_percentile(99);
+		print_percentile(99.9);
+		print_percentile(99.99);
 
 		info.clear();
 	}
+
+	void reportJSON(Stats & info, const std::string & filename)
+	{
+		std::ofstream jout(filename);
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Let use WriteBufferFromFile and DB/IO/Operators.h
By the way, it truncates existing file.

File: dbms/src/Client/Benchmark.cpp [hunk 2]
```diff
@@ -76,6 +78,11 @@ class Benchmark
 		else
 			throw Exception("Unknown query processing stage: " + stage, ErrorCodes::BAD_ARGUMENTS);
 
+		if (!json_path.empty() && Poco::File(json_path).exists()) /// Clear file with previous results
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: See below.

File: dbms/src/Client/Benchmark.cpp [hunk 3]
```diff
@@ -320,17 +340,60 @@ class Benchmark
...
+		};
 
-		std::cerr << "95%\t" 	<< info.sampler.quantileInterpolated(0.95) 	<< " sec.\n";
-		std::cerr << "99%\t" 	<< info.sampler.quantileInterpolated(0.99) 	<< " sec.\n";
-		std::cerr << "99.9%\t" 	<< info.sampler.quantileInterpolated(0.999) 	<< " sec.\n";
-		std::cerr << "99.99%\t" << info.sampler.quantileInterpolated(0.9999) << " sec.\n";
-		std::cerr << "100%\t" 	<< info.sampler.quantileInterpolated(1) 		<< " sec.\n";
+		for (int percent = 0; percent <= 90; percent += 10)
+			print_percentile(percent);
+		print_percentile(95);
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Empty line here will be helpful.

File: dbms/src/Client/Benchmark.cpp [hunk 4]
```diff
@@ -320,17 +340,60 @@ class Benchmark
...
+	{
+		std::ofstream jout(filename);
+		if (!jout.is_open())
+			throw Exception("Can't write JSON data");
+
+		std::lock_guard<std::mutex> lock(mutex);
+
+		double seconds = info.watch.elapsedSeconds();
+
+		jout << "{\n";
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: `jout` name is annoying for me, for unknown reason.
Let it be json_out.

File: dbms/src/Client/Benchmark.cpp [hunk 5]
```diff
@@ -183,18 +193,13 @@ class Benchmark
 		Stopwatch watch;
 
 		/// В цикле, кладём все запросы в очередь.
-		for (size_t i = 0; !interrupt_listener.check(); ++i)
+		for (size_t i = 0; !(max_iterations > 0) || i < max_iterations; ++i)
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Why not
`!max_iterations`
instead of
`!(max_iterations > 0)`
?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/src/Client/Benchmark.cpp (hunk 2 above)
Comment (ludv1x) [inline dbms/src/Client/Benchmark.cpp]: It is expected behavior. It allow to avoid reading data from previous test launches if current test is failed.

**Pull request merged**